File: examples/govern_demo.rs
//! Example: govern a task via `govern(config)` — `cargo run --example govern_demo`.
//! Mirrors the Python `with govern()` and REST `POST /govern` (one model, three surfaces).

use icebox::core::safety::{Charter, RiskLevel, ScopeManager};
use icebox::core::sdk::{govern, GovernanceConfig, GovernedOutcome, TaskSpec};
use icebox::core::{Capability, CvssScore, Role};
use serde_json::json;

#[tokio::main]
async fn main() {
    let rt = govern(GovernanceConfig {
        charter: Charter::accept("govern-demo", vec!["no destruction".into()]),
        scope: ScopeManager::new(vec!["10.0.0.0/24".into()]),
        max_risk: RiskLevel::Critical,
        role: Role::Admin,
        ..Default::default()
    });

    let task = TaskSpec {
        name: "scan".into(),
        target: "10.0.0.5".into(),
        capabilities: vec![Capability::NetworkScan],
        impact: RiskLevel::Low,
        destructive: false,
        options: [
            ("host".into(), "10.0.0.5".into()),
            ("ports".into(), "1-1024".into()),
        ]
        .into(),
        cvss: Some(CvssScore {
            cvss_v31: Some(9.5),
            cvss_v40: None,
            epss: Some(0.9),
            kev: true,
        }),
        ..Default::default()
    };

    let outcome: GovernedOutcome = rt
        .run(task, || async { Ok(json!({"open_ports": [22, 80, 443]})) })
        .await;

    match outcome {
        GovernedOutcome::Allowed {
            result,
            decision_id,
            ..
        } => {
            println!("Allowed (decision {decision_id}): {result}");
        }
        GovernedOutcome::Blocked {
            reason,
            decision_id,
        } => {
            println!("Blocked (decision {decision_id}): {reason}");
        }
        GovernedOutcome::NeedsApproval {
            reason,
            decision_id,
            approval_id,
        } => {
            println!("Needs approval {approval_id} (decision {decision_id}): {reason}");
        }
    }

    println!("Audit trail: {} records", rt.audit().await.len());
}

File: python/examples/govern_demo.py
"""Example: govern a task through ICEBOX using the flagship `govern()` API.

Mirrors the Rust `cargo run --example govern_demo` and the REST
`POST /govern` contract — one mental model across all three surfaces.

Requires a running ICEBOX daemon (`./target/release/icebox-daemon --api`):

    python examples/govern_demo.py
"""

from icebox import govern, Governance


# 1. The flagship single call — a governed session over HTTP.
#    `preflight` runs the action through the full GEE lifecycle; `complete`
#    records the outcome into the tamper-evident audit chain.
with govern() as g:
    verdict = g.preflight({
        "action": "scan",
        "target": "10.0.0.5",
        "capability": "network_scan",
        "impact": "low",
        "destructive": False,
    })
    print("preflight:", verdict)
    if verdict.get("approved"):
        # ... actually run your tool here ...
        outcome = {"success": True, "evidence": ["port 22 open"], "data": {}}
        print("complete:", g.complete(outcome, verdict.get("decision", "allow")))


# 2. In-process governance (no daemon) — same shape, native PyO3 runtime.
gov = Governance({
    "charter": {"accepted": True, "engagement": "demo", "rules_of_engagement": []},
    "scope": {"allow": ["10.0.0.0/24"]},
    "max_risk": "critical",
    "role": "admin",
})

result = gov.run({
    "name": "scan",
    "target": "10.0.0.5",
    "capabilities": ["network_scan"],
    "impact": "low",
    "destructive": False,
})
print("in-process run:", result)
